File: 0.1.1/activation_function/softmax.py
import numpy as np

def Function(inputs):
    return np.exp(inputs-max(inputs))/sum(np.exp(inputs-max(inputs)))

class Softmax:

    def function(Q, start_position, batch_size, value_matrix):
        output_matrix = np.zeros(np.shape(value_matrix))
        for i in range(start_position, start_position+batch_size):
            output_matrix[i] = Function(value_matrix[i])
        Q.put(output_matrix)

    def derivative(inputs, position):
        outputs = Function(inputs)
        derivative_vector = np.zeros(np.shape(inputs))
        for i in range(len(outputs)):
            if i!=position:
                derivative_vector[i] = -outputs[i]*outputs[position]
            else:
                derivative_vector[i] = outputs[i]*(1-outputs[i])
        return derivative_vector

File: 0.2.4/activation_function/softmax.jl
using .Threads

module Softmax
    function opt_func(values)
        return exp.(values.-maximum(values))/sum(exp.(values.-maximum(values)))
    end

    function func(value_matrix::Array{Float32})
        output_matrix = zeros(Float32, size(value_matrix))
        Threads.@threads for i in 1:size(value_matrix, 2)
            output_matrix[:,i] = opt_func(value_matrix[:,i])
        end
        return output_matrix
    end

    function diff(inputs::Array{Float32}, position::Int64)
        outputs = opt_func(inputs)
        derivative_vector = zeros(Float32, size(inputs))
        for i in 1:length(derivative_vector)
            derivative_vector[i] = i!=position ? -outputs[i]*outputs[position] : outputs[i]*(1-outputs[i])
        end
        return derivative_vector
    end

    function get_name()
        return "Softmax"
    end
end
